internal/log/log.go

```diff
@@ -0,0 +1,149 @@
+// Package log provides logging utilities for the tracer.
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Level specifies the logging level that the log package prints at.
+type Level int
+
+const (
+	// LevelDebug represents debug level messages.
+	LevelDebug Level = iota
+	// LevelWarn represents warning and errors.
+	LevelWarn
+)
+
+var prefixMsg = fmt.Sprintf("Datadog Tracer %s", version.Tag)
+
+var (
+	mu     sync.RWMutex   // guards below fields
+	level                 = LevelWarn
+	logger ddtrace.Logger = &defaultLogger{l: log.New(os.Stderr, "", log.LstdFlags)}
+)
+
+// UseLogger sets l as the active logger.
+func UseLogger(l ddtrace.Logger) {
+	mu.Lock()
+	defer mu.Unlock()
+	logger = l
+}
+
+// SetLevel sets the given lvl for logging.
+func SetLevel(lvl Level) {
+	mu.Lock()
+	defer mu.Unlock()
+	level = lvl
+}
+
+// Debug prints the given message if the level is LevelDebug.
+func Debug(fmt string, a ...interface{}) {
+	mu.RLock()
+	lvl := level
+	mu.RUnlock()
+	if lvl != LevelDebug {
+		return
+	}
+	printMsg("DEBUG", fmt, a...)
+}
+
+// Warn prints a warning message.
+func Warn(fmt string, a ...interface{}) {
+	printMsg("WARN", fmt, a...)
+}
+
+var (
+	errmu   sync.RWMutex                // guards below fields
+	erragg  = map[string]*errorReport{} // aggregated errors
+	errrate time.Duration               // the rate at which errors are reported
+	erron   bool                        // true if errors are being aggregated
+)
+
+func init() {
+	errrate = time.Minute
+	if v, ok := os.LookupEnv("DD_LOGGING_RATE"); ok {
+		if sec, err := strconv.ParseUint(v, 10, 64); err != nil {
+			Warn("Invalid value for DD_LOGGING_RATE: %v", err)
+		} else {
+			errrate = time.Duration(sec) * time.Second
+		}
+	}
+}
+
+type errorReport struct {
+	err   error
+	count uint64
+}
+
+// Error reports an error. Errors get aggregated and logged periodically. The
+// default is once per minute or once every DD_LOGGING_RATE number of seconds.
+func Error(format string, a ...interface{}) {
+	key := format // format should 99.9% of the time be constant
+	if reachedLimit(key) {
+		// avoid too much lock contention on spammy errors
+		return
+	}
+	errmu.Lock()
+	defer errmu.Unlock()
+	report, ok := erragg[key]
+	if !ok {
+		erragg[key] = &errorReport{err: fmt.Errorf(format, a...)}
```

Comment:
I think the first occurrence of an error needs to be reported, then additional ones counted and eventually flushed.
Assuming the logger emits timestamps, that gives at least one useful timestamp to look at when troubleshooting.

Reply:
I've implemented this based on our internal RFC:
> Tracer loggers should all default rate limiting to 1 unique message per 60 seconds.
This statement could be interpreted both like the current implementation, but also like your suggestion.

However, I think that doing it as you suggest will create a bad user experience and confusion because the user will see an error message and then the rest will be suppressed without a hint at the fact that there are more error messages being aggregated in the background. The core agent works like this however and they at least display a message, something like: "too many error messages, skipping for a bit...". I don't think we should do this. I think we should stay as we are.

However, looking at the RFC because of your comment I noticed that I missed two implementation details which I will add soon:
* The ability to disable this by setting `DD_LOGGING_RATE=0`
* Flushing all messages when the tracer stops.

Reply:
I did both of the points above and also added extra context into the error, showing when the first error occurred. I'm resolving but feel free to follow-up with a new comment thread if there's other things I might've missed.